File: backend/app/infrastructure/repositories/student_repository_impl.py
from typing import List, Optional
from sqlalchemy.orm import Session
from app.domain.entities.student import Alumno
from app.domain.repositories.student_repository import AlumnoRepository
from app.infrastructure.models.student_model import StudentModel
from app.infrastructure.repositories.course_repository_impl import (
    _grade_year_ordinal, _division_ordinal,
)


class AlumnoRepositoryImpl(AlumnoRepository):
    """Persists the frontend-facing `Alumno` DTO against `students`, which now
    has a direct `course_id` FK (no enrollment-history table)."""

    def __init__(self, db: Session):
        self.db = db

    def _to_entity(self, student: StudentModel) -> Alumno:
        course = student.course
        curso_label = ""
        if course is not None:
            curso_label = (
                f"{_grade_year_ordinal(course.grade_year)} "
                f"{_division_ordinal(course.division)} ({course.academic_year})"
            )

        return Alumno(
            id=str(student.id),
            nombre=student.first_name,
            apellido=student.last_name,
            dni=student.document_number or "",
            curso_id=str(student.course_id),
            curso=curso_label,
            recursante=student.is_repeating_student,
            porcentaje_asistencia=0.0,
            workshop_group_id=str(student.workshop_group_id) if student.workshop_group_id else None,
            taller=student.workshop_group.group_label if student.workshop_group else None,
        )

    def get_alumnos(self) -> List[Alumno]:
        rows = self.db.query(StudentModel).all()
        return [self._to_entity(s) for s in rows]

    def get_alumno_por_id(self, id: str) -> Optional[Alumno]:
        student = self.db.query(StudentModel).filter(StudentModel.id == id).first()
        return self._to_entity(student) if student else None

    def get_alumnos_por_curso(self, curso_id: str) -> List[Alumno]:
        rows = self.db.query(StudentModel).filter(StudentModel.course_id == curso_id).all()
        return [self._to_entity(s) for s in rows]

    def existe_dni(self, dni: str) -> bool:
        return (
            self.db.query(StudentModel)
            .filter(StudentModel.document_number == dni)
            .first()
            is not None
        )

    def crear_alumno_manual(
        self,
        first_name: str,
        last_name: str,
        national_id: str,
        course_id: str,
        workshop_group_id: Optional[str] = None,
    ) -> Alumno:
        student = StudentModel(
            first_name=first_name,
            last_name=last_name,
            document_number=national_id,
            course_id=course_id,
            workshop_group_id=workshop_group_id,
        )
        self.db.add(student)
        self.db.commit()
        self.db.refresh(student)
        return self._to_entity(student)

    def actualizar_alumno(self, id: str, alumno: Alumno) -> Optional[Alumno]:
        student = self.db.query(StudentModel).filter(StudentModel.id == id).first()
        if not student:
            return None

        student.first_name = alumno.nombre
        student.last_name = alumno.apellido
        student.document_number = alumno.dni
        if alumno.curso_id:
            student.course_id = alumno.curso_id
        student.workshop_group_id = alumno.workshop_group_id

        self.db.commit()
        self.db.refresh(student)
        return self._to_entity(student)

File: frontend/lib/infrastructure/repositories/student_repository_impl.dart
import '../../domain/entities/student.dart';
import '../../domain/entities/course.dart';
import '../../domain/entities/workshop_group.dart';
import '../../domain/repositories/student_repository.dart';
import '../datasources/api_datasource.dart';

class StudentRepositoryImpl implements StudentRepository {
  StudentRepositoryImpl(this._datasource);
  final ApiDatasource _datasource;

  @override
  Future<List<Student>> getAlumnos() => _datasource.getAlumnos();

  @override
  Future<List<Student>> getAlumnosPorCurso(String cursoId) async {
    final alumnos = await _datasource.getAlumnos();
    return alumnos.where((a) => a.cursoId == cursoId).toList();
  }

  @override
  Future<Student?> getAlumnoPorId(String id) async {
    final alumnos = await _datasource.getAlumnos();
    try {
      return alumnos.firstWhere((a) => a.id == id);
    } catch (_) {
      return null;
    }
  }

  @override
  Future<Student?> getAlumnoPorDni(String dni) async {
    final alumnos = await _datasource.getAlumnos();
    try {
      return alumnos.firstWhere((a) => a.dni == dni);
    } catch (_) {
      return null;
    }
  }

  @override
  Future<List<Course>> getCourses() => _datasource.getCursos();

  @override
  Future<List<WorkshopGroup>> getWorkshopGroupsByCourse(String courseId) =>
      _datasource.getWorkshopGroupsByCourse(courseId);

  @override
  Future<Student> crearAlumno({
    required String firstName,
    required String lastName,
    required String nationalId,
    required String courseId,
    String? workshopGroupId,
  }) =>
      _datasource.crearAlumno(
        firstName: firstName,
        lastName: lastName,
        nationalId: nationalId,
        courseId: courseId,
        workshopGroupId: workshopGroupId,
      );

  @override
  Future<Student> actualizarAlumno(Student alumno) => _datasource.actualizarAlumno(alumno);
}
